<launch>

    <!--For onboard computer visual-->
    <arg name="use_rviz_vins_fusion" default="true"/>
    <arg name="use_rviz_fast_planner" default="true"/>

    <!-- MAVROS configs -->
    <arg name="fcu_url" default="/dev/ttyACM0:57600" /> <!--Pixhawk usb'link port-->
	<arg name="gcs_url" default="" /> <!--QGC IP-->
    <!-- Fast-Planner configs -->
    <arg name="use_kino_planner" default="false" />
    <arg name="map_size_x" default="40.0"/>
    <arg name="map_size_y" default="40.0"/>
    <arg name="map_size_z" default=" 5.0"/>

    <arg name="odom_topic" default="/vins_estimator/odometry"/>
    <arg name="camera_pose_topic" default="/mavros/vision_pose/pose"/>
    <arg name="depth_topic" default="/camera/depth/image_rect_raw"/>

    <arg name="max_vel" default="1.0" />
    <arg name="max_acc" default="2.5" />

    <!-- Initial drone's position -->
    <arg name="init_pos_x" default="0.0" />
    <arg name="init_pos_y" default="0.0" />

    <!-- takeoff height -->
    <arg name="takeoff_height" default="1.0" />

    <!-- ROS static transformation -->
    <node pkg="tf" type="static_transform_publisher" name="world_map_linker" args="0 0 0 0 0 0 world map 100" />
    <node pkg="tf" type="static_transform_publisher" name="tf_camera" args="0.1 0 0.0 -1.57 0 -1.57 base_link camera_link 33"/>

    <!-- MAVROS + PX4 + RVIZ -->
    <include file="$(find experiment_fast_planner)/launch/mavros.launch" >
        <arg name="fcu_url" value="$(arg fcu_url)"/>
       
    </include>

    <!--  Camera -->
    <include file="$(find realsense2_camera)/launch/rs_d435i.launch" >
    </include>

    <!-- VINS-Fusion -->
    <include if="$(arg use_rviz_vins_fusion)" file="$(find vins)/launch/vins_rviz.launch" ></include>

    <!--include file="$(find experiment_fast_planner)/launch/vins_fusion.launch" ></include>-->

    <!-- VINS-Fusion Odem to Topic -->
    <node pkg="experiment_fast_planner" name="vins_transfer" type="vins_transfer.py" output="screen"/>

    <!-- Fast planner -->
    <include unless="$(arg use_kino_planner)" file="$(find experiment_fast_planner)/launch/px4_topo_replan.launch" >
        <arg name="map_size_x" default="$(arg map_size_x)"/>
        <arg name="map_size_y" default="$(arg map_size_y)"/>
        <arg name="map_size_z" default="$(arg map_size_z)"/>

        <arg name="odom_topic" default="$(arg odom_topic)" />
        <arg name="depth_topic" default="$(arg depth_topic)"/>
        <arg name="camera_pose_topic" default="$(arg camera_pose_topic)"/>
    </include>

    <include if="$(arg use_kino_planner)" file="$(find experiment_fast_planner)/launch/px4_kino_replan.launch" >
        <arg name="map_size_x" default="$(arg map_size_x)"/>
        <arg name="map_size_y" default="$(arg map_size_y)"/>
        <arg name="map_size_z" default="$(arg map_size_z)"/>

        <arg name="odom_topic" default="$(arg odom_topic)" />
        <arg name="depth_topic" default="$(arg depth_topic)"/>
        <arg name="camera_pose_topic" default="$(arg camera_pose_topic)"/>
    </include>


    <!-- Geometric controller -->
    <include file="$(find experiment_fast_planner)/launch/geometric_controller.launch">
        <arg name="max_acc" value="$(arg max_acc)" />
        <param name="Kp_x" value="6.0" />
        <param name="Kp_y" value="6.0" />
        <param name="Kp_z" value="10.0" />
        <param name="Kv_x" value="1.5" />
        <param name="Kv_y" value="1.5" />
        <param name="Kv_z" value="3.5" />
        <arg name="init_pos_x" value="0.0" />
        <arg name="init_pos_y" value="0.0" />
        <arg name="init_pos_z" value="$(arg takeoff_height)" />
    </include>

   

    <node pkg="experiment_fast_planner" name="traj_msg_converter" type="trajectory_msg_converter.py" output="screen">
        <param name="fast_planner_traj_topic" value="planning/ref_traj"/>
        <param name="traj_pub_topic" value="command/trajectory"/>
    </node>
    <!-- Rviz -->
    <include if="$(arg use_rviz_fast_planner)" file="$(find experiment_fast_planner)/launch/rviz.launch" />

</launch>
